lib/src/main/java/com/diffplug/spotless/shell/ShfmtStep.java

```diff
@@ -86,25 +86,40 @@ static class State implements Serializable {
 		// used for up-to-date checks and caching
 		final String version;
 		final transient ForeignExe exe;
-		// used for executing
-		private transient @Nullable List<String> args;
 
 		State(ShfmtStep step, ForeignExe pathToExe) {
 			this.version = step.version;
 			this.exe = Objects.requireNonNull(pathToExe);
 		}
 
+		FormatterFunc.Closeable toFunc() {
+			ProcessRunner runner = new ProcessRunner();
+			return FormatterFunc.Closeable.ofDangerous(runner, new ShfmtFilePathPassingFormatterFunc(runner, exe));
+		}
+	}
+
+	private static class ShfmtFilePathPassingFormatterFunc implements FormatterFunc.NeedsFile {
+		// used for executing
+		private transient @Nullable List<String> args;
+		final transient ForeignExe exe;
+		final transient ProcessRunner runner;
+
+		ShfmtFilePathPassingFormatterFunc(ProcessRunner runner, ForeignExe exe) {
+			this.runner = runner;
+			this.exe = exe;
+		}
+
+		@Override
+		public String applyWithFile(String unix, File file) throws Exception {
+			return format(runner, unix, file);
+		}
+
 		String format(ProcessRunner runner, String input, File file) throws IOException, InterruptedException {
 			if (args == null) {
-				args = List.of(exe.confirmVersionAndGetAbsolutePath(), "-i", "2", "-ci");
+				args = List.of(exe.confirmVersionAndGetAbsolutePath(), file.getAbsolutePath());
 			}
 
-			return runner.exec(input.getBytes(StandardCharsets.UTF_8), args).assertExitZero(StandardCharsets.UTF_8);
-		}
-
-		FormatterFunc.Closeable toFunc() {
-			ProcessRunner runner = new ProcessRunner();
-			return FormatterFunc.Closeable.of(runner, this::format);
+			return runner.exec(args).assertExitZero(StandardCharsets.UTF_8);
```

Comment:
While `shfmt` technically works when formatting text from standard input and respects the `.editorconfig`, I encountered issues making tests function correctly in this setup. Despite running a standalone test project successfully, executing `ShfmtTest` and `ShellExtensionTest` with a `.editorconfig` in the test project's directory did not yield the expected results.

To address this issue, I found success by providing the actual file path instead of standard input, allowing `shfmt` to properly respect the `.editorconfig` settings.

Reply:
This is no longer a relevant comment. I was able to get formatting to work with the input in standard in.